File: bubus-ts/examples/log_tree_demo.ts
import { z } from 'zod'

import { BaseEvent, EventBus } from '../src/index.js'

const RootEvent = BaseEvent.extend('RootEvent', {
  url: z.string(),
  event_result_schema: z.string(),
  event_result_type: 'string',
})

const ChildEvent = BaseEvent.extend('ChildEvent', {
  tab_id: z.string(),
  event_result_schema: z.string(),
  event_result_type: 'string',
})

const GrandchildEvent = BaseEvent.extend('GrandchildEvent', {
  status: z.string(),
  event_result_schema: z.string(),
  event_result_type: 'string',
})

const delay = (ms: number): Promise<void> =>
  new Promise((resolve) => {
    setTimeout(resolve, ms)
  })

async function main(): Promise<void> {
  const bus_a = new EventBus('BusA')
  const bus_b = new EventBus('BusB')

  async function forward_to_bus_b(event: InstanceType<typeof RootEvent>): Promise<string> {
    await delay(20)
    bus_b.dispatch(event)
    return 'forwarded_to_bus_b'
  }

  bus_a.on('*', forward_to_bus_b)

  async function root_fast_handler(event: InstanceType<typeof RootEvent>): Promise<string> {
    await delay(10)
    const child = event.bus?.emit(ChildEvent({ tab_id: 'tab-123', event_timeout: 0.1 }))
    if (child) {
      await child.done()
    }
    return 'root_fast_handler_ok'
  }

  async function root_slow_handler(event: InstanceType<typeof RootEvent>): Promise<string> {
    event.bus?.emit(ChildEvent({ tab_id: 'tab-timeout', event_timeout: 0.1 }))
    await delay(400)
    return 'root_slow_handler_timeout'
  }

  bus_a.on(RootEvent, root_fast_handler)
  bus_a.on(RootEvent, root_slow_handler)

  async function child_slow_handler(_event: InstanceType<typeof ChildEvent>): Promise<string> {
    await delay(150)
    return 'child_slow_handler_done'
  }

  async function child_fast_handler(event: InstanceType<typeof ChildEvent>): Promise<string> {
    await delay(10)
    const grandchild = event.bus?.emit(GrandchildEvent({ status: 'ok', event_timeout: 0.05 }))
    if (grandchild) {
      await grandchild.done()
    }
    return 'child_handler_ok'
  }

  async function grandchild_fast_handler(): Promise<string> {
    await delay(5)
    return 'grandchild_fast_handler_ok'
  }

  async function grandchild_slow_handler(): Promise<string> {
    await delay(60)
    return 'grandchild_slow_handler_timeout'
  }

  bus_b.on(ChildEvent, child_slow_handler)
  bus_b.on(ChildEvent, child_fast_handler)
  bus_b.on(GrandchildEvent, grandchild_fast_handler)
  bus_b.on(GrandchildEvent, grandchild_slow_handler)

  const root_event = bus_a.dispatch(RootEvent({ url: 'https://example.com', event_timeout: 0.25 }))

  await root_event.done()

  console.log('\n=== BusA logTree ===')
  console.log(bus_a.logTree())

  console.log('\n=== BusB logTree ===')
  console.log(bus_b.logTree())
}

await main()

File: examples/log_tree_demo.py
#!/usr/bin/env -S uv run python
"""Run: uv run python examples/log_tree_demo.py"""

import asyncio
from typing import Any

from bubus import BaseEvent, EventBus


class RootEvent(BaseEvent[str]):
    url: str


class ChildEvent(BaseEvent[str]):
    tab_id: str


class GrandchildEvent(BaseEvent[str]):
    status: str


async def delay_ms(ms: int) -> None:
    await asyncio.sleep(ms / 1000.0)


async def main() -> None:
    bus_a = EventBus('BusA')
    bus_b = EventBus('BusB')

    try:

        async def forward_to_bus_b(event: BaseEvent[Any]) -> str:
            await delay_ms(20)
            bus_b.emit(event)
            return 'forwarded_to_bus_b'

        bus_a.on('*', forward_to_bus_b)

        async def root_fast_handler(event: RootEvent) -> str:
            await delay_ms(10)
            child = event.event_bus.emit(ChildEvent(tab_id='tab-123', event_timeout=0.1))
            await child
            return 'root_fast_handler_ok'

        async def root_slow_handler(event: RootEvent) -> str:
            event.event_bus.emit(ChildEvent(tab_id='tab-timeout', event_timeout=0.1))
            await delay_ms(400)
            return 'root_slow_handler_timeout'

        bus_a.on(RootEvent, root_fast_handler)
        bus_a.on(RootEvent, root_slow_handler)

        async def child_slow_handler(_event: ChildEvent) -> str:
            await delay_ms(150)
            return 'child_slow_handler_done'

        async def child_fast_handler(event: ChildEvent) -> str:
            await delay_ms(10)
            grandchild = event.event_bus.emit(GrandchildEvent(status='ok', event_timeout=0.05))
            await grandchild
            return 'child_handler_ok'

        async def grandchild_fast_handler(_event: GrandchildEvent) -> str:
            await delay_ms(5)
            return 'grandchild_fast_handler_ok'

        async def grandchild_slow_handler(_event: GrandchildEvent) -> str:
            await delay_ms(60)
            return 'grandchild_slow_handler_timeout'

        bus_b.on(ChildEvent, child_slow_handler)
        bus_b.on(ChildEvent, child_fast_handler)
        bus_b.on(GrandchildEvent, grandchild_fast_handler)
        bus_b.on(GrandchildEvent, grandchild_slow_handler)

        root_event = bus_a.emit(RootEvent(url='https://example.com', event_timeout=0.25))
        await root_event

        print('\n=== BusA log_tree ===')
        print(bus_a.log_tree())

        print('\n=== BusB log_tree ===')
        print(bus_b.log_tree())
    finally:
        await bus_a.stop(clear=True, timeout=0)
        await bus_b.stop(clear=True, timeout=0)


if __name__ == '__main__':
    asyncio.run(main())
